server/app/services/rpc/node_service_pod_handler.rb

```diff
@@ -7,17 +7,25 @@ def initialize(grid)
       @lru_cache = LruRedux::ThreadSafeCache.new(1000)
     end
 
+    # @param service_instance [GridServiceInstance]
     def cached_pod(service_instance)
+      return nil unless service_instance.grid_service
+
       # TODO Do we need any other attrbutes for the composed key?
       cache_key = {
         id: service_instance.id,
         deploy_rev: service_instance.deploy_rev,
         desired_state: service_instance.desired_state
       }
-      @lru_cache.getset(cache_key) {
+      serializer = ServicePodSerializer.new(service_instance)
+      pod_hash = @lru_cache.getset(cache_key) {
         debug "pod_cache miss"
-        ServicePodSerializer.new(service_instance).to_hash if service_instance.grid_service
+        serializer.to_hash
       }
+      # hooks cannot be cached
+      pod_hash[:hooks] = serializer.build_hooks
```

Comment:
re-serializing the hooks for each service lookup is simpler, but doesn't it also end up being more expensive than the alternative policy of not caching serialized pods that contain oneshot hooks?

Like in https://github.com/kontena/kontena/pull/2701#discussion_r139705110 (although that writes to the LRU cache every time)

Reply:
Generating hooks should be lightweight because it does not fetch anything from the database (all data is already loaded).